### ISSUE TITLE
Payara Micro 5.2020.2: unable to inject EntityManager in async event /FISH-451

### ISSUE THREAD

**Issue by mlvtito** [payara/Payara#4815]
# Description #
----------

If I try to use EntityManager for the first time in an async CDI event handler, the EntityManager will never be injected in my app (even in synchronous threads). But if I use first the EntityManager in sync thread, everything seems to work fine (event if I use it in an async CDI event handler after that).

## Expected Outcome

I expect the EntityManager to be injected.
In my sample app, I expect to see this log:
```
####################### Processing event TestEvent[FIRE ASYNC]
```

## Current Outcome

This error occurs:
```
####################### Error while processing event
java.util.concurrent.CompletionException: java.lang.NullPointerException
	at org.jboss.weld.event.ObserverNotifier.handleExceptions(ObserverNotifier.java:463)
	at org.jboss.weld.event.ObserverNotifier.lambda$createSupplier$9(ObserverNotifier.java:451)
	at java.util.concurrent.CompletableFuture$AsyncSupply.run(CompletableFuture.java:1590)
	at org.glassfish.weld.services.ExecutorServicesImpl.lambda$inContextClassloader$1(ExecutorServicesImpl.java:122)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)
	Suppressed: java.lang.NullPointerException
		at net.rwx.issues.cdiasync.TestDAO.addData(TestDAO.java:25)
		at net.rwx.issues.cdiasync.TestDAO$Proxy$_$$_WeldClientProxy.addData(Unknown Source)
		at net.rwx.issues.cdiasync.TestService.observeAsyncEvent(TestService.java:34)
		at net.rwx.issues.cdiasync.TestService$Proxy$_$$_WeldSubclass.observeAsyncEvent$$super(Unknown Source)
		at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
		at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
		at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
		at java.lang.reflect.Method.invoke(Method.java:498)
		at org.jboss.weld.interceptor.proxy.TerminalAroundInvokeInvocationContext.proceedInternal(TerminalAroundInvokeInvocationContext.java:51)
		at org.jboss.weld.interceptor.proxy.AroundInvokeInvocationContext.proceed(AroundInvokeInvocationContext.java:78)
		at org.glassfish.jersey.ext.cdi1x.transaction.internal.WebAppExceptionInterceptor.intercept(WebAppExceptionInterceptor.java:53)
		at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
		at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
		at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
		at java.lang.reflect.Method.invoke(Method.java:498)
		at org.jboss.weld.interceptor.reader.SimpleInterceptorInvocation$SimpleMethodInvocation.invoke(SimpleInterceptorInvocation.java:73)
		at org.jboss.weld.interceptor.proxy.NonTerminalAroundInvokeInvocationContext.proceedInternal(NonTerminalAroundInvokeInvocationContext.java:66)
		at org.jboss.weld.interceptor.proxy.AroundInvokeInvocationContext.proceed(AroundInvokeInvocationContext.java:78)
		at org.glassfish.cdi.transaction.TransactionalInterceptorBase.proceed(TransactionalInterceptorBase.java:212)
		at org.glassfish.cdi.transaction.TransactionalInterceptorRequired.transactional(TransactionalInterceptorRequired.java:115)
		at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
		at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
		at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
		at java.lang.reflect.Method.invoke(Method.java:498)
		at org.jboss.weld.interceptor.reader.SimpleInterceptorInvocation$SimpleMethodInvocation.invoke(SimpleInterceptorInvocation.java:73)
		at org.jboss.weld.interceptor.proxy.InterceptorMethodHandler.executeAroundInvoke(InterceptorMethodHandler.java:84)
		at org.jboss.weld.interceptor.proxy.InterceptorMethodHandler.executeInterception(InterceptorMethodHandler.java:72)
		at org.jboss.weld.interceptor.proxy.InterceptorMethodHandler.invoke(InterceptorMethodHandler.java:56)
		at org.jboss.weld.bean.proxy.CombinedInterceptorAndDecoratorStackMethodHandler.invoke(CombinedInterceptorAndDecoratorStackMethodHandler.java:79)
		at org.jboss.weld.bean.proxy.CombinedInterceptorAndDecoratorStackMethodHandler.invoke(CombinedInterceptorAndDecoratorStackMethodHandler.java:68)
		at net.rwx.issues.cdiasync.TestService$Proxy$_$$_WeldSubclass.observeAsyncEvent(Unknown Source)
		at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
		at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
		at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
		at java.lang.reflect.Method.invoke(Method.java:498)
		at org.jboss.weld.injection.StaticMethodInjectionPoint.invoke(StaticMethodInjectionPoint.java:95)
		at org.jboss.weld.injection.StaticMethodInjectionPoint.invoke(StaticMethodInjectionPoint.java:85)
		at org.jboss.weld.injection.MethodInvocationStrategy$DefaultMethodInvocationStrategy.invoke(MethodInvocationStrategy.java:144)
		at org.jboss.weld.event.ObserverMethodImpl.sendEvent(ObserverMethodImpl.java:330)
		at org.jboss.weld.event.ObserverMethodImpl.sendEvent(ObserverMethodImpl.java:308)
		at org.jboss.weld.event.ObserverMethodImpl.notify(ObserverMethodImpl.java:286)
		at javax.enterprise.inject.spi.ObserverMethod.notify(ObserverMethod.java:124)
		at org.jboss.weld.util.Observers.notify(Observers.java:166)
		at org.jboss.weld.event.ObserverNotifier.notifyAsyncObserver(ObserverNotifier.java:415)
		at org.jboss.weld.event.ObserverNotifier.lambda$notifyAsyncObservers$5(ObserverNotifier.java:358)
		at org.jboss.weld.event.ObserverNotifier.lambda$null$8(ObserverNotifier.java:442)
		at org.jboss.weld.security.spi.SecurityServices.lambda$getSecurityContextAssociator$0(SecurityServices.java:80)
		at org.jboss.weld.event.ObserverNotifier.lambda$createSupplier$9(ObserverNotifier.java:438)
		... 5 more
	[CIRCULAR REFERENCE:java.lang.NullPointerException]
]]
```

## Steps to reproduce (Only for bug reports) 

1 -**  clone the sample project
```
git clone https://github.com/mlvtito/issue-entity-manager.git
```

2 -**  build the project
```
mvn clean install
```

3 -**  start payara micro
```
mvn payara-micro:start
```

4 -**  invoke async event first (to see the NullPointerException)
```
http://localhost:8080/issue-cdi-async/api/test/async
```
then
```
http://localhost:8080/issue-cdi-async/api/test/sync
```

5 -**  stop and restart payara micro

6 -**  invoke sync event first (everything is working fine)
```
http://localhost:8080/issue-cdi-async/api/test/sync
```
then
```
http://localhost:8080/issue-cdi-async/api/test/async
```
	
### Samples

https://github.com/mlvtito/issue-entity-manager

By changing the property `version.payara.micro` in `pom.xml`, I noticed that the error appears for the firt time in version 5.192. Before that, there was another bug with CDI async event. I have to go back to version 5.183 to have everything alright.

## Environment ##

- **Payara Version**: 5.2020.2
- **Edition**: Micro
- **JDK Version**: OpenJDK Runtime Environment (build 1.8.0_252-8u252-b09-1ubuntu1-b09)
- **Operating System**: Linux Ubuntu 20.04
- **Database**: No Database

**Comment by lprimak** [comment]
try adding ```@Transactional``` to ```net.rwx.issues.cdiasync.TestDAO```

**Comment by mlvtito** [comment]
I've just tried to add @Transactional on TestDAO class and method but the problem is still here. When I add this annotation, I see another warning in the log file.

So it means that I see in the log:
* the NullPointerException
* this warning (which, I guess, is related to @Transactional on TestService): ``TransactionalInterceptorBase.markThreadAsTransactional currentInvocation==null
* and this one which appeared with the added annotation on TestDAO: ``No ComponentInvocation present for @Transactional annotation processing. Restriction on use of UserTransaction will not be enforced.

**Comment by lprimak** [comment]
Well, at least that'll help diagnose the problem, thanks!

**Comment by rdebusscher** [comment]
Hi Arnaud,

Thank you for the reproducer of this issue. I have created the internal ticket `FISH-451` to investigate this issue related to Weld.  When the class `TestDAO` is marked with `@Stateless` instead of _@ApplicationScoped_, the injection is performed correctly.

Best Regards
Rudy

**Comment by mlvtito** [comment]
Hello Rudy,

Thanks for the tip. I'll use this annotation until this issue is resolved.

Best regards

**Comment by sgflt** [comment]
Related StackOverflow https://stackoverflow.com/questions/49339263/java-ee-8-no-transaction-in-asynchronous-observer-cdi-2-0

**Comment by thehpi** [comment]
If you use a managed executor it also works oke

    @Resource
    ManagedExecutorService threadPool;

    public void produceAsyncEvent(String data) {
        events.fireAsync(TestEvent.of("FIRE ASYNC"), NotificationOptions.ofExecutor(threadPool)).exceptionally((t) -> {
            logger.log(Level.SEVERE, t, () -> "####################### Error while processing event");
            return null;
        });
    }

**Comment by thehpi** [comment]
This issue is fixed in version 5.2021.9. So I guess it can be closed.

**Comment by rdebusscher** [comment]
Hi Hans,

Thx for mentioning, it seems this issue was solved as a side effect of the upgrade of Weld we did for JKD 17 support. https://github.com/payara/Payara/pull/5484

Thx
Rudy
